### PR TITLE
store lastmodified field in accountdata

### PR DESCRIPTION
**Pull request by vogel**
All other EntryFrame types (account, offer, trust) store value of
lastModifiedLedgerSeq in database in lastmodified field. Now data also
stores that.

Lack of this field caused errors in stellar-core-acceptance-history-
testnet-catchup: EntryFrame::checkAgainstDatabase was loading data
from database but with lastModifiedLedgerSeq set to 0 - it was causing
"Inconsistent state between objects" exception.

Signed-off-by: Rafał Malinowski <rafal.przemyslaw.malinowski@gmail.com>

### CODE REVIEW

**Comment by vogel** [comment]
I'm not exactly sure if we should update database from 4->5 in the way that is implemented in this PR. I did it that way because that is how all other updates are done, but to me it seems like we are losing a lot of valuable data in this case.

Should it be changed to ALTER TABLE to add the new row and just ignore values of lastmodified for old items? I don't think we are checking them in other places than in "checkdb" and there it won't care if old values have 0 or valid data here.

**Comment by vogel** [comment]
Changed database update code

**Comment by vogel** [comment]
Changed target branch

File: src/database/Database.cpp [hunk 1]
```diff
@@ -119,6 +119,7 @@ Database::applySchemaUpgrade(unsigned long vers)
 
     case 4:
         BanManager::dropAll(*this);
+        DataFrame::dropAll(*this); // not and error, it changed again
```
Comment (MonsieurNicolas) [inline src/database/Database.cpp]: this change is actually wrong: deleting from the database doesn't delete from the bucket list so I think this will cause entries to show up as "create" instead of "updates" and potentially cause weird problems down the line with shadow entries. The proper way to do this is to do a alter table where lastmodified is something consistent with the past (I think it was stored as 0 in the xdr with the previous bug)

Comment (MonsieurNicolas) [inline src/database/Database.cpp]: you also need to rev SCHEMA_VERSION to perform the update correctly

Comment (jedmccaleb) [inline src/database/Database.cpp]: damn I didn't see lastmodified in the xdr. This is a feature flag then since the older ledgers need to keep lastmodified as 0. 

SCHEMA_VERSION was rev'd. we are bundling several changes into this branch

Comment (jedmccaleb) [inline src/database/Database.cpp]: Speaking of lastmodified, do you remember the main motivation? was it just for horizon?

Comment (vogel) [inline src/database/Database.cpp]: I'll check if the lastmodified fiels was stored as '0' in XDR files.
It is possible that it was only zeroed in database, but this still needs to be checked.

Comment (MonsieurNicolas) [inline src/database/Database.cpp]: lastmodified is used for a couple things:
* meta data generation
* history traversal (allows to skip to the right ledger to build the history for that ledger entry)

the problem with the drop table is more that we "leak" entries in the bucket list, and with that potentially generate the wrong deltas (which will cause the ledger hashes to differ between machines and therefore invalidate the ledger chain)

Comment (vogel) [inline src/database/Database.cpp]: Short version: this change affects only `txmeta` field of `txhistory` and fixes a serious crash in `PARANOID_MODE`.

Long version:

I`m not sure if we should guard this change with a feature flag. It only affects `txmeta` field of `txhistory` table in database (details below).

Also, with current implementation it is possible to easily crash stellar-core - it is enough to send a transaction to network that changes or
removes one `DATA` ledger entry (but this requires `PARANOID_MODE=true` in configuration).

I`ll start with `PARANOID_MODE` set to true. It causes method `LedgerDelta::checkAgainstDatabase()` to be really executed. It calls
`EntryFrame::checkAgainstDatabase()` which for DATA entries load entry with `lastModifiedLedgerSeq` set by default to 0, as this value
is currently not stored in database. It makes following code to throw an exception:

	void
	EntryFrame::checkAgainstDatabase(LedgerEntry const& entry, Database& db)
	{
		auto key = LedgerEntryKey(entry);
		flushCachedEntry(key, db);
		auto const& fromDb = EntryFrame::storeLoad(key, db);
		if (!(fromDb->mEntry == entry))
		{
			std::string s;
			s = "Inconsistent state between objects: ";
			s += xdr::xdr_to_string(fromDb->mEntry, "db");
			s += xdr::xdr_to_string(entry, "live");
			throw std::runtime_error(s);
		}
	}

It can be fixed with workaround - by setting `fromDb->mEntry->lastModifiedLedgerSeq` to `entry.lastModifiedLedgerSeq` when type is `DATA`,
`fromDb->mEntry->lastModifiedLedgerSeq` is zero and currnent protocol version is <= 2.

I would like to mention here that this bug was not catched by our unit tests. `ManageDataTests.cpp` contains tests for adding, changing and
modyfing `DATA` ledger entries, it even runs with `PARANOID_MODE=true`. But unfortunately, it never closes the ledger which causes
`checkAgainstDatabase` to be called. So the `PARANOID_MODE=true` has no effects on these tests at all. I think that either `applyCheck()`
should close a ledger with only one transaction (I`ve tried that, but there are some issues to be resolved), or at least `LedgerDelta::checkAgainstDatabase()`
should be called manually around `app.getLedgerManager().checkDbState();` call (tried that one as well, works good and finds the bug mentioned
above). I`ll be happy with either change, but closing ledger seems to test behavior matching live node more closely.

Lets get back to `lastModifiedLedgerSeq` value for `DATA` ledger entries.

There are two methods of storing LedgerEntry datatypes. First one is XDR streams. In this case there is no problem in storing/restoring `lastModifiedLedgerSeq` value for all LedgerEntry types (including `DATA`) - xdrpp takes care of that.

Second method is using database. LedgerEntry instances are stored in four tables:
* `ACCOUNT` - table `accounts` WITH field `lastmodified`
* `TRUSTLINE` - table `trustlines` WITH field `lastmodified`
* `OFFER` - table `offers` WITH field `lastmodified`
* `DATA` - table `accountdata` WITHOUT field `lastmodified`

So we can see that ACCOUNT, TRUSTLINE and OFFER behavior did not change in any way.

Lets look as DATA LedgerEntry type (represented by `DataEntry` class in C++ code) and how it can be restored from database. It is only done inside `DataFrame` class in two methods:

`DataFrame::loadData()`
`DataFrame::loadAccountsData()`
`DataFrame::loadAllData()`

All three methods set `lastModifiedLedgerSeq` of `LedgerEntry` to 0 instead of proper value because it is not stored in database at the moment.

* `loadAllData()` in called by LedgerManagerImpl::checkDbState() which only check key and count - invalid `lastModifiedLedgerSeq` value does not matter
* `loadAccountsData()` is only called in `TxTests.cpp` - so we can ignore this one

There are different call sites for `loadData()`:

* `applyManageData()` from `TxTests.cpp` - so we can ignore this one (this even does not read `lastModifiedLedgerSeq` value)
* `EntryFrame::storeLoad()` - this one is called by `EntryFrame::checkAgainstDatabase()` - so this is cause of crash when `PARANOID_MODE=true` - this PR fixes that 

`ManageDataOpFrame::doApply()` is more interesting. This is the only place where `DATA` loaded from database is stored somewhere else.
Lets examin that store. In two places `delta.recordEntry()` is called. So it gets stored in `mPrevious` field of `LedgerDelta`. From there it could only
get to `mPrevious` of different `LedgerDelta` instance (in `mergeEntries()`) or it can be saved to `LedgerEntryChanges` in `LedgerDelta::addCurrentMeta`
(called by `LedgerDelta::getChanges()`.

The last method is called in:
* `LedgerDeltaTests.cpp` - can be ignored
* `TxTests.cpp` - can be ignored
* `LedgerManagerImpl::processFeesSeqNums()` - we can ignore it as well, because the delta that call `getChanges()` will only contain `ACCOUNT` entries
  (ensured by `TransactionFrame::processFeeSeqNum()`)
* `TransactionFrame::apply()` is the last place where it is called

Following `apply()` method (version that takes `TransactionMeta` as secon argument) we get:
* `TxTests.cpp` - can be ignored
* `LedgerManagerImpl::applyTransactions()`

And the final step of this flow analysis: `TransactionMeta` that can contain invalid value of `lastModifiedLedgerSeq` is finally passed to `TransactionFrame::storeTransaction` where
that invalid meta value is stored under `txmeta` field of `txhistory` table.

If value of this fields must be the same over all instances of `stellar-core` over the world then this change should be guarded by feature flag, but then
we need workaround for fixing crash mentioned at the begining of this discussion.

Another concern is *will it change the way hashes are created* and the answer is no.
Here is list of hashes that we use:

* `SCPNomination::quorumSetHash` - computed from `SCPQuorumSet`, nothing to do with `DATA`
* `SCPStatement::quorumSetHash` - computed from `SCPQuorumSet`, nothing to do with `DATA`
* `SCPStatement::commitQuorumSetHash` - computed from `SCPQuorumSet`, nothing to do with `DATA`
* `StellarValue::txSetHash` - computed from `TxSetFrame::getContentsHash()`, uses `previousLedgerHash` and `TransactionEnvelope` to hash, neither of them contains any
  reference to `DateEntry` structure (it works on `Operations` as in `ManageDataOp` not on `LedgerEntry` as `DataEntry`)
* `LedgerHeader::previousLedgerHash` - copied from `LedgerHeaderHistoryEntry::hash`
* `LedgerHeader::txSetResultHash` - computed from `TransactionResultSet` which again, does not have any reference to `DataEntry`
* `LedgerHeader::bucketListHash` - will be discussed below
* `LedgerHeader::skipList[4]` - copied from `LedgerHeader::bucketListHash`
* `TransactionSet::previousLedgerHash` - copied from `LedgerHeader::previousLedgerHash`
* `TransactionResultPair::transactionHash` - computed from `TransactionFrame::getContentsHash()`, works on `Operations as in `ManageDataOp` not on `LedgerEntry` as `DataEntry`
* `LedgerHeaderHistoryEntry::hash` - copied from `LedgerHeaderFrame::mHash`, computed from `LedgerHeader` - which is safe
* `Hello::networkID` - computed from network passphase
* `Memo::hash` - unused in stellar-core
* `Memo::retHash` - unused in stellar-core
* `TransactionSignaturePayload::networkId` - computed from network passphase

So the only one interesting hash is `bucketListHash`. It is computed in few layers, but on bottom is `Bucket::OutputIterator::getBucket`.
It createsh hash for Bucked in `Bucket::fresh` method - exactly from data stored into bucket file. This data is in form of two types of
entries `BucketEntry` - `LIVEENTRY` represented by `LedgerEntry` (this one can be dangerous) and `DEADENTRY` respresnted by `LedgerKey`
(this one is safe because it does not contain `lastModifiedLedgerSeq` value).

We can follow where `liveEntries` are coming from, and it is `BucketList::addBatch()` called by `LedgerManagerImpl::closeLedgerHelper`
with `delta.getLiveEntries()` which in turn returns only data stored in `mNew` and `mMod` fields. And we saw above that the only place
in `LedgerDelta` where values with invalid `lastModifiedLedgerSeq` can be stored is `mPrevious`.




So, in summary:
* applying this patch wont change value of any bucket, ledger or transaction
* applying this patch wont affect history in any way (as it uses buckets)
* applying this patch will only change values of `txmeta` field in `txhistory` table in database - if this is enough to make this a feature flag, then I'll implement this with workaround for crash
* applying this patch will fix crash in `PARANOID_MODE`
